documentation/Calls.md

```diff
@@ -0,0 +1,636 @@
+   (15.4.2.4(21)); `INTENT()` can't always be checked.
+1. Non-simply-contiguous (9.5.4) arrays being passed to non-`POINTER`
+   dummy arguments that must be contiguous due to a `CONTIGUOUS`
+   attribute or not being assumed-shape or assumed-rank (which
+   is always the case for F77ish procedures).
+   This should be a runtime decision, so that effective arguments
+   that turn out to be contiguous can be passed cheaply.
+   This rule does not apply to coarray dummies, whose effective arguments
+   are required to be simply contiguous when this rule would otherwise
+   force the use of a temporary (15.5.2.8); neither does it apply
+   to `ASYNCHRONOUS` and `VOLATILE` effective arguments, which are
+   disallowed when it matters (C1539, C1540).
+   *Only temporaries created by this contiguity requirement are
+   candidates for being copied back to the original variable after
+   the call* (see below).
+
+Fortran requires (18.3.6(5)) that interoperable procedure dummy
+arguments with `CONTIGUOUS` &/or assumed-length `CHARACTER(*)`
+explicit-shape/assumed-size arrays handle the compaction of
+discontiguous data *in the callee*, at least when called from C.
+And discontiguous data must be compacted on the *caller's* side
+when passed from Fortran to C (18.3.6(6)).
+
+We could perform all argument compaction (discretionary or
+required) in the callee, but there are many cases where the
+compiler knows that the effective argument data are contiguous
+when compiling the caller (a temporary is needed for other reasons,
+or the effective argument is simply contiguous) and a run-time test for
+discontiguity in the callee can be avoided by using a caller-compaction
+convention when we have the freedom to choose.
+
+While we are unlikely to want to _needlessly_ use a temporary for
+an effective argument that does not require one for any of these
+reasons above, we are specifically disallowed from doing so
+by the standard in cases where pointers to the original target
+data are required to be valid across the call (15.5.2.4(9-10)).
+In particular, compaction of assumed-shape arrays for discretionary
+contiguity on the leading dimension to ease SIMD vectorization
+cannot be done safely for `TARGET` dummies without `VALUE`.
+
+Effective arguments associated with known `INTENT(OUT)` dummies that
+require allocation of a temporary -- and this can only be for reasons of
+contiguity -- don't have to populate it, but they do have to perform
+minimal initialization of any `ALLOCATABLE` components so that
+the runtime doesn't crash when the callee finalizes and deallocates
+them.
+Note that calls to implicit interfaces must conservatively allow
+for the use of `INTENT(OUT)` by the callee.
+
+Except for `VALUE` and known `INTENT(IN)` dummy arguments, the original
+contents of local designators that have been compacted into temporaries
+could optionally have their `ALLOCATABLE` components invalidated
+across the call as an aid to debugging.
```

Comment:
I do not understand what is meant by "invalidated" here. Does this mean to de-allocate the `ALLOCATABLE` components of the original content that was compacted ? Is it to ensure they are not accessed inside the callee ?

Reply:
It's sounds worth testing. I guess we may want to come with a clever "merry-go-round" of allocatable component storages anyway to save some memory allocation and de-allocation around calls. Instead of copying allocatable components we could "move" them into the temp and move them back into the original variable after the call.